src/data/set/intervals/instances.lean

```diff
@@ -0,0 +1,261 @@
+{ mul := λ p q, ⟨p*q, ⟨mul_nonneg p.2.1 q.2.1, mul_le_one p.2.2 q.2.1 q.2.2⟩⟩ }
+
+instance has_pow : has_pow (Icc (0:α) 1) ℕ :=
+{ pow := λ p n, ⟨p.1 ^ n, ⟨pow_nonneg p.2.1 n, pow_le_one n p.2.1 p.2.2⟩⟩ }
+
+@[simp, norm_cast] lemma coe_mul (x y : Icc (0:α) 1) : ↑(x * y) = (x * y : α) := rfl
+@[simp, norm_cast] lemma coe_pow (x : Icc (0:α) 1) (n : ℕ) : ↑(x ^ n) = (x ^ n : α) := rfl
+
+lemma mul_le_left {x y : Icc (0:α) 1} : x * y ≤ x :=
+(mul_le_mul_of_nonneg_left y.2.2 x.2.1).trans_eq (mul_one x)
+
+lemma mul_le_right {x y : Icc (0:α) 1} : x * y ≤ y :=
+(mul_le_mul_of_nonneg_right x.2.2 y.2.1).trans_eq (one_mul y)
+
+instance monoid_with_zero : monoid_with_zero (Icc (0:α) 1) :=
+subtype.coe_injective.monoid_with_zero _ coe_zero coe_one coe_mul coe_pow
+
+instance comm_monoid_with_zero {α : Type*} [ordered_comm_semiring α] :
+  comm_monoid_with_zero (Icc (0:α) 1) :=
+subtype.coe_injective.comm_monoid_with_zero _ coe_zero coe_one coe_mul coe_pow
+
+instance cancel_monoid_with_zero {α : Type*} [ordered_ring α] [no_zero_divisors α] :
+  cancel_monoid_with_zero (Icc (0:α) 1) :=
+@function.injective.cancel_monoid_with_zero α _ no_zero_divisors.to_cancel_monoid_with_zero
+  _ _ _ _ coe subtype.coe_injective coe_zero coe_one coe_mul coe_pow
+
+instance cancel_comm_monoid_with_zero {α : Type*} [ordered_comm_ring α] [no_zero_divisors α] :
+  cancel_comm_monoid_with_zero (Icc (0:α) 1) :=
+@function.injective.cancel_comm_monoid_with_zero α _
+  no_zero_divisors.to_cancel_comm_monoid_with_zero
+  _ _ _ _ coe subtype.coe_injective coe_zero coe_one coe_mul coe_pow
+
+variables {β : Type*} [ordered_ring β]
+
+lemma one_sub_mem {t : β} (ht : t ∈ Icc (0:β) 1) : 1 - t ∈ Icc (0:β) 1 :=
+by { rw mem_Icc at *, exact ⟨sub_nonneg.2 ht.2, (sub_le_self_iff _).2 ht.1⟩ }
+
+lemma mem_iff_one_sub_mem {t : β} : t ∈ Icc (0:β) 1 ↔ 1 - t ∈ Icc (0:β) 1 :=
+⟨one_sub_mem, λ h, (sub_sub_cancel 1 t) ▸ one_sub_mem h⟩
+
+lemma one_minus_nonneg (x : Icc (0:β) 1) : 0 ≤ 1 - (x : β) := by simpa using x.2.2
+lemma one_minus_le_one (x : Icc (0:β) 1) : 1 - (x : β) ≤ 1 := by simpa using x.2.1
+
+variables {R : Type*} [linear_ordered_field R]
+
+lemma div_mem  {x y : R} (hx : 0 ≤ x) (hy : 0 ≤ y) (hxy : x ≤ y) : x / y ∈ Icc (0:R) 1 :=
+⟨div_nonneg hx hy, div_le_one_of_le hxy hy⟩
+
+lemma mul_pos_mem_iff {a t : R} (ha : 0 < a) : a * t ∈ Icc (0:R) 1 ↔ t ∈ set.Icc (0:R) (1/a) :=
+begin
+  split; rintros ⟨h₁, h₂⟩; split,
+  { exact nonneg_of_mul_nonneg_left h₁ ha },
+  { rwa [le_div_iff ha, mul_comm] },
+  { exact mul_nonneg ha.le h₁ },
+  { rwa [le_div_iff ha, mul_comm] at h₂ }
+end
+
+lemma two_mul_sub_one_mem_iff {t : R} : 2 * t - 1 ∈ Icc (0:R) 1 ↔ t ∈ set.Icc (1/2 : R) 1 :=
+by split; rintros ⟨h₁, h₂⟩; split; linarith
+
+@[simp] lemma proj_Icc_eq_zero {x : R} : proj_Icc (0 : R) 1 zero_le_one x = 0 ↔ x ≤ 0 :=
+proj_Icc_eq_left zero_lt_one
+
+@[simp] lemma proj_Icc_eq_one {x : R} : proj_Icc (0 : R) 1 zero_le_one x = 1 ↔ 1 ≤ x :=
+proj_Icc_eq_right zero_lt_one
+
+end set.Icc
```

Comment:
This doesn't look like it belongs in the `set.Icc` namespace since `set.Icc` doesn't appear anywhere in the expression:
```suggestion
end set.Icc

namespace set

@[simp] lemma proj_Icc_eq_zero {x : R} : proj_Icc (0 : R) 1 zero_le_one x = 0 ↔ x ≤ 0 :=
proj_Icc_eq_left zero_lt_one

@[simp] lemma proj_Icc_eq_one {x : R} : proj_Icc (0 : R) 1 zero_le_one x = 1 ↔ 1 ≤ x :=
proj_Icc_eq_right zero_lt_one

end set
```